### ISSUE TITLE
Cannot create an anonymous cart on Reaction and Starter Kit simultaneously

### ISSUE THREAD

**Issue by mikemurray** [reactioncommerce/reaction#4686]
## Issue Description

If you create an anonymous cart from the starter-kit and try to add items to an anonymous cart on reaction, you'll get an error. The same is true in reverse. This also happens across browsers, so between Firefox and Chrome I'm seeing the same issue.

reaction_1            | Exception while invoking method 'cart/createCart' { MongoError: E11000 duplicate key error collection: reaction.Cart index: c2_userId dup key: { : null }
reaction_1            |     at Function.MongoError.create (/home/node/.meteor/packages/npm-mongo/.3.1.1-rc18.14.1w24u71.2w11++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/error.js:49:10)
reaction_1            |     at toError (/home/node/.meteor/packages/npm-mongo/.3.1.1-rc18.14.1w24u71.2w11++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:149:22)
reaction_1            |     at coll.s.topology.insert (/home/node/.meteor/packages/npm-mongo/.3.1.1-rc18.14.1w24u71.2w11++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/operations/collection_ops.js:803:39)
reaction_1            |     at /home/node/.meteor/packages/npm-mongo/.3.1.1-rc18.14.1w24u71.2w11++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/connection/pool.js:531:18
reaction_1            |     at _combinedTickCallback (internal/process/next_tick.js:131:7)
reaction_1            |     at process._tickDomainCallback (internal/process/next_tick.js:218:9)
reaction_1            |  => awaited here:
reaction_1            |     at Function.Promise.await (/home/node/.meteor/packages/promise/.0.11.1.1vczms8.mxi5++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:56:12)
reaction_1            |     at Promise.asyncApply (imports/plugins/core/cart/server/no-meteor/mutations/createCart.js:103:9)
reaction_1            |     at /home/node/.meteor/packages/promise/.0.11.1.1vczms8.mxi5++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/fiber_pool.js:43:40
reaction_1            |  => awaited here:
reaction_1            |     at Function.Promise.await (/home/node/.meteor/packages/promise/.0.11.1.1vczms8.mxi5++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:56:12)
reaction_1            |     at MethodInvocation.createCartMethod (imports/plugins/core/cart/server/methods/createCart.js:31:18)
reaction_1            |     at currentArgumentChecker.withValue (packages/check/match.js:118:15)
reaction_1            |     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
reaction_1            |     at Object._failIfArgumentsAreNotAllChecked (packages/check/match.js:116:43)
reaction_1            |     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1764:18)
reaction_1            |     at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:719:19)
reaction_1            |     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
reaction_1            |     at DDPServer._CurrentWriteFence.withValue (packages/ddp-server/livedata_server.js:717:46)
reaction_1            |     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
reaction_1            |     at Promise (packages/ddp-server/livedata_server.js:715:46)
reaction_1            |     at new Promise (<anonymous>)
reaction_1            |     at Session.method (packages/ddp-server/livedata_server.js:689:23)
reaction_1            |     at packages/ddp-server/livedata_server.js:559:43
reaction_1            |   name: 'MongoError',
reaction_1            |   message: 'E11000 duplicate key error collection: reaction.Cart index: c2_userId dup key: { : null }',
reaction_1            |   driver: true,
reaction_1            |   index: 0,
reaction_1            |   code: 11000,
reaction_1            |   errmsg: 'E11000 duplicate key error collection: reaction.Cart index: c2_userId dup key: { : null }',
reaction_1            |   [Symbol(mongoErrorContextSymbol)]: {} }

## Possible Solution

Seems `userId` is set to `null`, and we are probably indexing on `userId`. We might be able solve this with a randomly generated userId, or base it on the token if the schema allows it.

## Steps to Reproduce

1. As an anonymous user
1. Add items to the cart from the starter kit to create a cart
1. On reaction, try to add items to the cart, again as an anonymous user
1. You should receive an error

### Versions

Node: 10.7.0
NPM: 6.2.0
Meteor Node: 8.11.4
Meteor NPM: 6.4.1
Reaction CLI: 0.29.0
Reaction: 2.0.0
Reaction branch: master
Docker: 18.06.0-ce

**Comment by aldeed** [comment]
We can just omit `userId` in the database rather than setting it to `null`.

**Comment by rosshadden** [comment]
This is happening because of a holdover index: `Cart.c2_user`. Migration `30` drops this index, but because as developers we often switch around between branches and releases, the index gets re-added. And it does not get removed again, as we do not rollback migrations.

It might be a good idea to think about us applying rollback migrations on startup if the codebase's latest migration doesn't match the one in the database. Let's keep that in mind for if something like this happens again.

In summary, this is only an issue in our dev workflows, and should very rarely be seen in the wild.

**Comment by willmoss1000** [comment]
Sorry to comment on this closed issue, but I'm getting this error on a production site, and I'm still on 1.4.2. I've check the database and we don't have an index on `userId` in the cart collection. How can I resolve this?

Full trace:
```
MongoError: E11000 duplicate key error index: hopexchange.Cart.$c2_userId dup key: { : null }
    at Function.MongoError.create (/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/error.js:31:11)
    at /app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:497:72
    at authenticateStragglers (/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:443:16)
    at Connection.messageHandler (/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:477:5)
    at Socket.<anonymous> (/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:333:22)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:594:20)
```

**Comment by zenweasel** [comment]
@willmoss1000 Are you using the starterkit? Probably not if you are on 1.4.2 (are you sure that's your version number? That's a pre-marketplace version) 

If not, not sure this issue is related to what you are seeing?

**Comment by willmoss1000** [comment]
@zenweasel sorry! **1.4.1** -  I'm not using Starterkit, however I'm experience the exact same error as far as I can see? That's why I felt it was worth posting here?

**Comment by aldeed** [comment]
@willmoss1000 No disrespect intended, but it seems unlikely that MongoDB would throw an index error if there isn't an index.

Connect to db with a Mongo shell and triple check that it's the same DB that your app instance with the error is connected to. Then run `db.Cart.getIndexes()` and look for one with the name `c2_userId`. That will be the one to delete.

Note that connecting an old app version to that db will automatically re-create that index on startup and require you to manually delete it again.

**Comment by willmoss1000** [comment]
@aldeed thanks for this. Yes, the index is not there when I connect to the Mongo shell and this is certainly the right DB instance. However, I'm not seeing this error anymore weirdly.

I haven't deployed since thenso I can't understand why this has gone away. Perhaps mongo is smart enough to automatically remove an index if it can't be established?

Anyway - thanks for your help.
